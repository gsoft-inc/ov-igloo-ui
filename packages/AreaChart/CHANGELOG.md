# @igloo-ui/area-chart

## 1.0.4

### Patch Changes

- 7d36815: Updated hopper token versions and igloo icon versions.

## 1.0.3

### Patch Changes

- a2df55d: Updated to latest Hopper Token dependency

## 1.0.2

### Patch Changes

- b99b307: Removed the error caused when all the scores in the AreaChart are null.

## 1.0.1

### Patch Changes

- 9739155: Updated storybook and recharts. Applied the new brand look for filter.
- 9739155: Gave ColorPicker the new Wokleap look. Color now uses css variables to update its color. Fixed spacing issues in tooltip. Updated hopper token version.

## 1.0.0

### Major Changes

- 4096378: Applied the new brand style to AreaChart and removed the gradient area under the line.

  ## BREAKING CHANGE

  The 'withColoredArea' prop was removed from AreaChart. This is because we removed the gradient area completely.

## 0.6.3

### Patch Changes

- bf4ddef: Updated storybook version and linting. Made linting fixes in these components.

## 0.6.2

### Patch Changes

- f07b40f: Fixed the AreaChart styling for the empty data message. It is also responsive now.

## 0.6.1

### Patch Changes

- 89d10c5: Updated recharts version and fixed dropdown mock

## 0.6.0

### Minor Changes

- a1bcd80: Add dotted line area for null values

## 0.5.2

### Patch Changes

- 56c75d0: Fixed the y-axis width calculation hook to work properly when usign the loading prop.

## 0.5.1

### Patch Changes

- a41e1d4: Fix: moved destructuring of props to function parameters so that storybook will automatically show the default value in the props table. It was done in Modal because it needed to send props to useDialog. In this case, @default annotation was used in comments.

## 0.5.0

### Minor Changes

- 589b6ea: displays in the x-axis + tooltip the utc date instead of the local date

## 0.4.0

### Minor Changes

- 77e4f12: Added special month formatting for May, June, July for English locale only in function formatSpecialMonth
  Added new props for setting the maximum elements of the x-axis

## 0.3.0

### Minor Changes

- 1b4cf68: Added a locale prop to be able to change languages

## 0.2.1

### Patch Changes

- 93dd0cb: fix(AreaChart): Issues with loading and empty state
  - rename the css animation to avoid conflicts
  - update y-axis size when the loading state disappeared
  - change the orientation of animation for empty state

## 0.2.0

### Minor Changes

- c4eb19a: add a loading state

## 0.1.2

### Patch Changes

- 0855a3c: The extra space ont he left of the area chart was removed

## 0.1.1

### Patch Changes

- 6ee9c45: Updated anything causing errors or warnings in the browser console

## 0.1.0

### Minor Changes

- 83a9e99: Feat: initial release of the area chart component
