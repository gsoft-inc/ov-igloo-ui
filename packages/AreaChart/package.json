{
  "name": "@igloo-ui/area-chart",
  "version": "1.0.4",
  "main": "dist/AreaChart.js",
  "module": "dist/AreaChart.js",
  "types": "dist/AreaChart.d.ts",
  "styles": "dist/area-chart.css",
  "publishConfig": {
    "access": "public"
  },
  "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
  "author": "Officevibe Inc.",
  "contributors": [
    "Vicky Comeau vicky.comeau@gsoft.com"
  ],
  "license": "Apache-2.0",
  "peerDependencies": {
    "react": ">= 16.8.6"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js"
  },
  "dependencies": {
    "@igloo-ui/tokens": "^2.0.0",
    "@hopper-ui/tokens": "^3.1.2",
    "classnames": "^2.3.2",
    "luxon": "^3.2.1",
    "recharts": "^2.9.2"
  },
  "devDependencies": {
    "@types/recharts": "^1.8.26"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "IE >= 11"
  ]
}
