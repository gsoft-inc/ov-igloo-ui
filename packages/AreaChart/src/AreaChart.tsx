import * as React from 'react';
import cx from 'classnames';
import { DateTime, Settings } from 'luxon';
import {
  AreaChart as RechartsAreaChart,
  Area,
  XAxis,
  YAxis,
  Tooltip,
  ResponsiveContainer,
  CartesianGrid,
  XAxisProps,
  YAxisProps,
  DotProps,
  ReferenceArea,
} from 'recharts';
import { CurveType } from 'recharts/types/shape/Curve';

import variables from '@igloo-ui/tokens/dist/base10/tokens.json';

import ChartTooltip from './ChartTooltip';
import './area-chart.scss';
import useDynamicYAxisWidth from './hooks/useDynamicYAxisWidth';

interface DataSet {
  /** Date/time in ISO format */
  dateTimeStamp: string;
  /** The score displayed on the y axis */
  score: number | null;
  /** The secondary score that is displayed inside the tooltip */
  secondaryScore?: number;
  /** The text displayed beside the score in the tooltip */
  name?: string;
  /** The text displayed beside the secondary score in the tooltip */
  secondaryName?: string;
}

interface AreaChartData {
  /** Date/time displayed as a number */
  dateTimeStamp: number;
  /** The score displayed on the y axis */
  score: number | null;
  /** The secondary score that is displayed inside the tooltip */
  secondaryScore?: number;
  /** The text displayed beside the score in the tooltip */
  name?: string;
  /** The text displayed beside the secondary score in the tooltip */
  secondaryName?: string;
}

interface DataRange {
  /** The min value of the y axis score
   * (Possible values: number, 'auto', 'dataMin' or 'dataMax')
   */
  min: number | string;
  /** The min value of the y axis score
   * (Possible values: number, 'auto', 'dataMin' or 'dataMax')
   */
  max: number | string;
}

interface DateTimeRange {
  /** Start date/time in ISO format */
  start: string;
  /** End date/time in ISO format */
  end: string;
}

export interface AreaChartProps extends React.ComponentProps<'div'> {
  /** Add a specific class to the area chart component */
  className?: string;
  /** The data object that will be used for the area chart */
  dataSet: DataSet[];
  /** Add a data-test tag for automated tests */
  dataTest?: string;
  /** The start date and end date for the x axis */
  dateRange: DateTimeRange;
  /** Whether or not the chart should resize */
  isResponsive?: boolean;
  /** Update the default locale */
  locale?: string;
  /** The min and max value of the y-axis
   * (Possible values: number, 'auto', 'dataMin' or 'dataMax')
   */
  range?: DataRange;
  /** The maximum elements of the x-axis */
  xAxisTickCount?: number;
  /** The formatter function for the y-axis */
  scoreFormatter?: (score: number) => string;
  /** The formatter function for the score on the tooltip */
  tooltipScoreFormatter?: (score: number) => string;
  /** The message that is displayed when data is unavailable */
  unavailableDataMessage?: string;
  /** Whether or not to show the colored area below the line */
  withColoredArea?: boolean;
  /** Replaces AreaChart label with a skeleton */
  loading?: boolean;
}

const AreaChart: React.FunctionComponent<AreaChartProps> = ({
  className,
  dataSet,
  dataTest,
  dateRange,
  loading = false,
  isResponsive = true,
  locale = 'en',
  range = { max: 'auto', min: 0 },
  xAxisTickCount = 7,
  scoreFormatter,
  tooltipScoreFormatter,
  unavailableDataMessage,
  withColoredArea = true,
}: AreaChartProps) => {
  const DEFAULT_SKELETON_WIDTH = 24;
  const DEFAULT_SKELETON_HEIGHT = 8;

  Settings.defaultLocale = locale;

  const SkeletonAxisTick = ({
    x,
    y,
    className,
    payload,
    skeletonWidth = DEFAULT_SKELETON_WIDTH,
    skeletonHeight = DEFAULT_SKELETON_HEIGHT,
    orientation,
  }: {
    x?: number;
    y?: number;
    className?: string;
    payload?: { offset: number };
    skeletonWidth?: number;
    skeletonHeight?: number;
    orientation?: string;
  }): React.ReactElement => {
    let positionX;
    let positionY;

    if (x && y && payload) {
      const positionBottomCenter = x - skeletonWidth / 2 + payload.offset / 2;
      const positionLeft = x - skeletonWidth;

      positionX =
        orientation === 'bottom' ? positionBottomCenter : positionLeft;
      positionY = y - skeletonHeight / 2 + payload.offset / 2;
    }

    return (
      <g>
        <rect
          x={positionX}
          y={positionY}
          className={cx('ids-area-chart-skeleton-animation', className)}
          width={skeletonWidth}
          height={skeletonHeight}
          rx="4"
        />
      </g>
    );
  };

  const classes = cx('ids-area-chart', className);

  const startDate = DateTime.fromISO(dateRange.start, { zone: 'utc' })
    .endOf('day')
    .toUTC();
  const endDate = DateTime.fromISO(dateRange.end, { zone: 'utc' })
    .endOf('day')
    .toUTC();

  const [areaChartData, setAreaChartData] = React.useState<AreaChartData[]>();

  const formatSpecialMonth = (date: DateTime): string => {
    if (date.locale !== 'en') {
      return date.toLocaleString({
        month: 'short',
        day: 'numeric',
        timeZone: 'utc',
      });
    }

    const utcDate = date.toUTC();
    const { month } = utcDate;
    let monthString = '';
    switch (month) {
      case 5:
        monthString = 'May';
        break;
      case 6:
        monthString = 'June';
        break;
      case 7:
        monthString = 'July';
        break;
      case 9:
        monthString = 'Sept.';
        break;
      default:
        monthString = date.toFormat('MMM.');
    }
    return `${monthString} ${utcDate.day}`;
  };

  const dateFormatter = (date: number): string => {
    const newDatefromMillis = DateTime.fromMillis(date);
    return formatSpecialMonth(newDatefromMillis);
  };

  const tooltipDateFormatter = (date: number): string => {
    return DateTime.fromMillis(date).toLocaleString({
      year: 'numeric',
      month: 'long',
      day: 'numeric',
      timeZone: 'utc',
    });
  };

  const getNumberOfTicks = (linePoints: number, maxTicks: number): number => {
    let bestTick = 0;
    let bestScore = 0;

    for (let i = maxTicks; i > 2; i -= 1) {
      const score = ((linePoints - i) / (i - 1)) % 1;
      if (score === 0) {
        bestTick = i;
        break;
      }

      if (!bestScore || score > bestScore) {
        bestTick = i;
        bestScore = score;
      }
    }

    return bestTick;
  };

  const getTicks = (): number[] => {
    const numberOfDays = endDate.diff(startDate, ['days']).days + 1;
    const numberOfTicks = getNumberOfTicks(numberOfDays, xAxisTickCount);
    const ticksArr = [];
    const daysBetween = Math.ceil(numberOfDays / numberOfTicks);
    let currentDate = endDate;

    ticksArr.push(endDate.valueOf());
    for (let i = numberOfTicks; i > 0; i -= 1) {
      currentDate = currentDate.minus({ days: daysBetween });
      if (currentDate.valueOf() <= startDate.valueOf()) {
        break;
      }
      ticksArr.push(currentDate.valueOf());
    }

    ticksArr.push(startDate.valueOf());

    return ticksArr;
  };

  const geYtTicks = (): number[] => {
    let ticks: number[] = [];
    if (
      range &&
      typeof range.min === 'number' &&
      typeof range.max === 'number'
    ) {
      ticks = Array.from(
        { length: range.max + 1 },
        (_, i) => i + (range.min as number)
      );
    } else {
      ticks = [1, 2, 3, 4, 5];
    }

    return ticks;
  };

  const { yAxisWidth, setChartRef } = useDynamicYAxisWidth({
    yAxisWidthModifier: (x) => x + 20,
    loading,
    areaChartData,
  });

  const cartesianGridConfig = {
    stroke: variables.grey200,
    strokeOpacity: 1,
    vertical: false,
  };

  const tickStyle = {
    fill: variables.grey600,
    fontSize: variables.fontSize2,
    fillOpacity: 1,
  };

  const xAxisConfig: XAxisProps = {
    dataKey: 'dateTimeStamp',
    scale: 'time',
    tickFormatter: dateFormatter,
    type: 'number',
    domain: [startDate.valueOf(), endDate.valueOf()],
    interval: 'preserveStartEnd',
    ticks: getTicks(),
    tick: tickStyle,
    tickLine: false,
    minTickGap: 20,
    tickMargin: 10,
    allowDataOverflow: true,
    axisLine: {
      stroke: variables.grey400,
    },
  };

  const yAxisConfig: YAxisProps = {
    allowDecimals: true,
    type: 'number',
    tickFormatter: scoreFormatter,
    tickLine: false,
    tickMargin: 10,
    tick: tickStyle,
    domain: [range.min, range.max],
    axisLine: {
      stroke: variables.grey400,
    },
    width: yAxisWidth,
  };

  if (!dataSet.length) {
    yAxisConfig.ticks = geYtTicks();
  }

  if (loading) {
    yAxisConfig.ticks = geYtTicks();
    yAxisConfig.tick = <SkeletonAxisTick />;
    xAxisConfig.tick = <SkeletonAxisTick skeletonWidth={32} />;
  }

  const dotConfig: DotProps = {
    r: 4,
    strokeWidth: 1,
    stroke: '#FFF',
    fill: variables.electricBlue500,
  };

  const commonAreaConfig = {
    type: 'monotone' as CurveType,
    strokeWidth: 3,
    fill: withColoredArea ? 'url(#curveAreaFillGradient)' : 'none',
    fillOpacity: 1,
  };

  const areaConfig = {
    ...commonAreaConfig,
    dataKey: 'score',
    connectNulls: false,
    stroke: variables.electricBlue500,
    activeDot: {
      ...dotConfig,
    },
  };

  const unavailableDataConfig = {
    ...commonAreaConfig,
    dataKey: 'score',
    connectNulls: true,
    stroke: variables.grey400,
    strokeWidth: '6',
    strokeDasharray: '1,9',
    activeDot: false,
  };

  const onlyUnavailableDataConfig = {
    type: 'basis',
    stroke: 'none',
    fill: 'none',
  };

  const buildAreaDefs = (): JSX.Element => {
    return (
      <defs>
        <linearGradient id="curveAreaFillGradient" x1="0" y1="0" x2="0" y2="1">
          <stop stopColor={variables.electricBlue50} />
          <stop offset="78%" stopColor="#EDF6FF" />
          <stop offset="96%" stopColor={variables.samoyed} />
        </linearGradient>
      </defs>
    );
  };

  React.useEffect(() => {
    let updatedAreaChartData = [];

    if (dataSet.length) {
      updatedAreaChartData = dataSet.map((dataSet) => {
        const currentScore: number | null = dataSet.score;

        return {
          dateTimeStamp: DateTime.fromISO(dataSet.dateTimeStamp)
            .toUTC()
            .endOf('day')
            .valueOf(),
          score: currentScore,
          secondaryScore: dataSet.secondaryScore,
          name: dataSet.name,
          secondaryName: dataSet.secondaryName,
        };
      });
    } else {
      const dates = getTicks();
      updatedAreaChartData = dates
        .map((date) => {
          return {
            dateTimeStamp: date,
            score: 0,
          };
        })
        .reverse();
    }

    setAreaChartData(updatedAreaChartData);
    // If getTicks() is added to dependency array, it creates an infinite loop
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [dataSet]);

  const areaChart = (
    <RechartsAreaChart
      data={areaChartData}
      margin={{ right: 26, top: 10, bottom: 10, left: 0 }}
      ref={setChartRef}
    >
      {withColoredArea && buildAreaDefs()}
      <CartesianGrid {...cartesianGridConfig} />
      <XAxis {...xAxisConfig} />
      <YAxis {...yAxisConfig} />
      {dataSet.length && !loading ? (
        <Area {...areaConfig} />
      ) : (
        <>
          <ReferenceArea
            x1={startDate.valueOf()}
            x2={endDate.valueOf()}
            {...onlyUnavailableDataConfig}
            label={unavailableDataMessage}
          />
          {!loading && (
            <Area {...unavailableDataConfig} strokeLinecap="round" />
          )}
        </>
      )}
      {dataSet.length ? (
        <Tooltip
          filterNull={false}
          cursor={false}
          content={
            <ChartTooltip
              dateFormatter={tooltipDateFormatter}
              scoreFormatter={tooltipScoreFormatter}
              unavailableDataMessage={unavailableDataMessage}
            />
          }
        />
      ) : (
        ''
      )}
    </RechartsAreaChart>
  );

  return (
    <div className={classes} data-test={dataTest}>
      {isResponsive ? (
        <ResponsiveContainer width="100%" height={220}>
          {areaChart}
        </ResponsiveContainer>
      ) : (
        areaChart
      )}
    </div>
  );
};

export default AreaChart;
