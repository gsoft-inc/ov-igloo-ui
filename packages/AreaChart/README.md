# AreaChart

The area chart is a line chart with shading below the line. This area chart depicts score over time.

<Example />

<ReferenceLinks />

## Installation

To install `@igloo-ui/area-chart` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/area-chart
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/area-chart
```

## Usage

Then to use the component in your code just import it!

```jsx
import AreaChart from '@igloo-ui/area-chart';

const nFormatter = (num: number): string => {
  const lookup = [
    { value: 1, symbol: '' },
    { value: 1e3, symbol: 'k' },
    { value: 1e6, symbol: 'M' },
    { value: 1e9, symbol: 'B' },
    { value: 1e12, symbol: 'T' },
  ];
  const digits = 2;
  const rx = /\.0+$|(\.[0-9]*[1-9])0+$/;
  var item = lookup
    .slice()
    .reverse()
    .find(function (item) {
      return num >= item.value;
    });
  return item
    ? (num / item.value).toFixed(digits).replace(rx, '$1') + item.symbol
    : '0';
};

<AreaChart
  scoreFormatter={nFormatter}
  dateRange={{
    start: '2022-10-01T17:40:40.457Z',
    end: '2022-10-07T17:40:40.457Z',
  }}
  dataSet={[
    {
      dateTimeStamp: '2022-10-01T00:40:40.457Z',
      score: 48878,
      name: 'sent',
      secondaryName: 'sent',
    },
    {
      dateTimeStamp: '2022-10-02',
      score: 49879,
      name: 'sent',
      secondaryName: 'sent',
    },
    {
      dateTimeStamp: '2022-10-03',
      score: 33587,
      name: 'sent',
      secondaryName: 'sent',
    },
    {
      dateTimeStamp: '2022-10-04',
      score: 45445,
      name: 'sent',
      secondaryName: 'sent',
    },
    {
      dateTimeStamp: '2022-10-05',
      score: 34534,
      secondaryScore: 30000,
      name: 'sent',
      secondaryName: 'sent',
    },
    {
      dateTimeStamp: '2022-10-06',
      score: 34555,
      secondaryScore: 52677,
      name: 'sent',
      secondaryName: 'sent',
    },
    {
      dateTimeStamp: '2022-10-07',
      score: 67897,
      name: 'sent',
      secondaryName: 'sent',
    },
  ]}
/>;
```
