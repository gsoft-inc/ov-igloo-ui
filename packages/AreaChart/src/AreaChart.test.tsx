/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, screen } from '@testing-library/react';
import IglooProvider from "@igloo-ui/provider";

import AreaChart from './AreaChart';

describe('AreaChart', () => {
  test('It should render without error and a snapshot', () => {
    const { asFragment } = render(
      <IglooProvider locale="en-US">
        <AreaChart
          scoreFormatter={(score) => `${score}%`}
          isResponsive={false}
          dataTest="areaChart1"
          dateRange={{
            start: '2022-10-01T17:40:40.457Z',
            end: '2022-10-07T17:40:40.457Z',
          }}
          dataSet={[
            {
              dateTimeStamp: '2022-10-01T00:40:40.457Z',
              score: 48878,
              name: 'sent',
              secondaryName: 'sent',
            },
            {
              dateTimeStamp: '2022-10-02',
              score: 49879,
              name: 'sent',
              secondaryName: 'sent',
            },
            {
              dateTimeStamp: '2022-10-03',
              score: 33587,
              name: 'sent',
              secondaryName: 'sent',
            },
            {
              dateTimeStamp: '2022-10-04',
              score: 45445,
              name: 'sent',
              secondaryName: 'sent',
            },
            {
              dateTimeStamp: '2022-10-05',
              score: 34534,
              secondaryScore: 30000,
              name: 'sent',
              secondaryName: 'sent',
            },
            {
              dateTimeStamp: '2022-10-06',
              score: 34555,
              secondaryScore: 52677,
              name: 'sent',
              secondaryName: 'sent',
            },
            {
              dateTimeStamp: '2022-10-07',
              score: 67897,
              name: 'sent',
              secondaryName: 'sent',
            },
          ]}
        />
      </IglooProvider>
    );
    expect(screen.getByTestId('areaChart1')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
