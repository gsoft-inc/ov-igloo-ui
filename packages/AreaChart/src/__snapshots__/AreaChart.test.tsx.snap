// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AreaChart It should render without error and a snapshot 1`] = `
<DocumentFragment>
  <div
    class="ids-area-chart"
    data-test="areaChart1"
  />
</DocumentFragment>
`;
